server/controllers/awsController.js

```diff
@@ -0,0 +1,74 @@
+import regeneratorRuntime from "regenerator-runtime";
+require('dotenv').config();
+const AWS = require('aws-sdk');
+const accessKeyId = process.env.AWS_ACCESS_KEY;
+const secretAccessKey = process.env.AWS_SECRET_KEY;
+const bucketName = process.env.BUCKET_NAME;
+const bucketRegion = process.env.BUCKET_REGION;
+const SIZE_OF_ASSETID = 66;
+export let filesByAssetId = {};
+const s3bucket = new AWS.S3({
+  accessKeyId,
+  secretAccessKey,
+  region: bucketRegion,
+  Bucket: bucketName,
+});
+
+export const handleFileUpload = (file, assetId) => {
+  new Promise(async (resolve, reject) => {
+    const responseData = [];
+    file.map((item) => {
+      const params = {
+        Bucket: bucketName,
+        Key: `${assetId}:${item.originalname}`,
+        Body: item.buffer,
+        ACL: 'public-read',
+      };
+      s3bucket.upload(params, (err, data) => {
+        if (err) {
+          reject();
+        } else {
+          responseData.push(data);
+          if (responseData.length === file.length) {
+            ProcessFilesForAssets();
```

Comment:
From my understanding of how `ProcessFilesForAssets` works, it will retry as many times as it can until it success. However, you are already resolving the promise of `handleFileUpload` when you don't know if `ProcessFilesForAssets` has been completed.

Why not wrap `ProcessFilesForAssets` as a Promise and then only solve it if things worked out?

Reply:
That is because it is not imperative 'ProcessFilesForAssets' goes through at the same time 'handleFileUpload' does or in the same call/moment. We want to resolve 'handleFileUpload' asap since the client is waiting for an answer. Knowing the files have been uploaded is what we care most about in this case.

'ProcessFilesForAssets' is only the method that gives Go an object containing what files the different assetIDs have, also important it stays up to date but not as much as making sure we resolve that call. An argument could be made against my view though, so depending on the requirements we can change the code. I feel like its fine as it is since 'ProcessFilesForAssets' is almost certain to always work (it is AWS afterall) but if for some reason it doesn't, we still want to make sure we resolve the call as soon as the files have been uploaded.

Reply:
Understood, fair enough.